### ISSUE TITLE
git-clang-format: add return codes for better pipeline usage

### ISSUE THREAD

**Issue by Febbe** [llvm/llvm-project#53220]
It would be useful, if git-clang-format could return a returncode != 0, when 
 - git-clang-format is in the `--diff` mode **and**
 - there are files to format

This would make it possible, to let a pipeline step fail, without writing a script around git-clang-format to interpret the output.
Currently, anyone has to check for the output string, if it contains something like `index 78cf3fe..71f38cf 100644` or `no modified files to format`. This is error-prone and a subject of failure, when a string changes.

**Comment by mkurdej** [comment]
Review: https://reviews.llvm.org/D117414.

**Comment by owenca** [comment]
Fixed in edbb8a8.

**Comment by jarmar** [comment]
I believe the last line of the script needs to be modified, too, to say `sys.exit(main())` instead of just `main()`. Currently, the script always returns 0 for me.

**Comment by Febbe** [comment]
@jarmar did you checked that you are using this version of the script, the fix is relatively new and might not be in llvm14. Also you could check if the return code changes for you, when you apply that change. But in python, the last value in a function is implicitly the return value, so it should make no difference.

**Comment by jarmar** [comment]
Yes, I downloaded `git-clang-format` directly from the main branch. I don't believe your remark about implicit return values in Python is correct, but either way, the issue here is with the return value of the script itself (rather than a function).

**Comment by Febbe** [comment]
> @jarmar did you checked that you are using this version of the script, the fix is relatively new and might not be in llvm14. Also you could check if the return code changes for you, when you apply that change. But in python, the last value in a function is implicitly the return value, so it should make no difference.

Ok, obviously I am not familiar enough with python. There is no main function and returning a value from the top level is just ignored or a syntax error. So there is definitely a sys.exit call missing.

**Comment by owenca** [comment]
> I believe the last line of the script needs to be modified, too, to say `sys.exit(main())` instead of just `main()`. Currently, the script always returns 0 for me.

Can you create a new issue and describe your use case or how to reproduce it?

**Comment by touraill-adobe** [comment]
This change probably broke most scripts using `--diff`, it would have been nice if the new behavior had been conditioned by an option (like `git diff --exit-code` for instance). Not only we need to update scripts but it is also quite hard to write them so that they are compatible with both old and new versions of `git-clang-format`.

**Comment by Febbe** [comment]
At the time of the change, yes, I agree but that change is pretty old now and the respective LLVM tool chain also. Why do you need to support so outdated LLVM toolchains?

**Comment by touraill-adobe** [comment]
First of all, updating toolchains is not always as straightforward as you would think, there are many projects out there still building with mscv 2017, gcc 7 or clang 10. 2 years is not that long ago :slightly_smiling_face: 

The fact that we are only hitting this issue now shows that developers themselves don't update their toolchain all the time (rightfully so, as it can be downright incompatible with their project, or disrupt their work by requiring changes here and there to support the new version).

In the case of `clang-format` there is an even greater incentive to _not_ upgrade because versions are not at all backward compatible. Formatting code with `clang-format-n` doesn't produce the same result as formatting it with `clang-format-n+1`, with the same config file. 

Since we don't want to reformat our entire code base every two months, we stick to a given version for a long time, so most developers keep this old version. But we also have developers that build with `clang`, with a more recent version than what we require for `clang-format`. Thankfully, `git-clang-format` can be configured to use a specific `clang-format` binary, so it is possible to have a recent LLVM in the path while still using an older `clang-format`. However this means that we have developers that are using the old `git-clang-format` (because they just need the old `clang-format`) and developers that are using the new `git-clang-format` (because they want a recent `clang` in their `PATH`).

So I don't _need_ to support "outdated" versions of `git-clang-format`, but if I decide to support only the new ones I need to either ask my teammates to install the new one alongside the old one, or ask my teammates to install the new one and reformat our entire codebase (which is something we will do eventually, but I wish we would do it because we want to and not because we are forced to :sweat_smile:). And if I decide to stick with the old `git-clang-format` for now I need to ask my teammates with newer LLVM in their `PATH` to tweak it to use the `git-clang-format` from their older LLVM.

Note that I'm not asking for any change today: as you said, the breaking change has already been done, several months ago, so modifying the behavior again would be even worse; I just wanted to underline the fact that breaking interfaces of command line tools can be a burden for users, to hopefully make it happen less in the future.

**Comment by Febbe** [comment]
> if I decide to support only the new ones I need to either ask my teammates to install the new one alongside the old one, or ask my teammates to install the new one and reformat our entire codebase (which is something we will do eventually, but I wish we would do it because we want to and not because we are forced to 😅)

What we do at work is to never reformat the entire codebase, just rely on git-clang-format to only reformat changed lines of code. 
We always use the latest llvm toolchain. That's an approach not feasible for open source projects, since you can't enforce every contributor to always install the latest toolchain, without loosing not so sophisticated developers. That's the reason xournalpp for example allows the latest 3 clang-format versions. So the code must be either formatted with 17,16 or 15.

> Note that I'm not asking for any change today: as you said, the breaking change has already been done, several months ago, so modifying the behavior again would be even worse; I just wanted to underline the fact that breaking interfaces of command line tools can be a burden for users, to hopefully make it happen less in the future.

Yes, I fully agree and understand this, I even proposed this in the feature request I made. Glad for me, that I don't care, when newer versions of a software will break things as long it's still an improvement :D.
